File: examples/contrib/nontransitive_dice.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Linq;
using System.Text.RegularExpressions;
using Google.OrTools.ConstraintSolver;

public class NonTransitiveDice
{
    /**
     *
     * Nontransitive dice.
     *
     * From
     * http://en.wikipedia.org/wiki/Nontransitive_dice
     * """
     * A set of nontransitive dice is a set of dice for which the relation
     * 'is more likely to roll a higher number' is not transitive. See also
     * intransitivity.
     *
     * This situation is similar to that in the game Rock, Paper, Scissors,
     * in which each element has an advantage over one choice and a
     * disadvantage to the other.
     * """
     *
     * Also see http://www.hakank.org/or-tools/nontransitive_dice.py
     *
     *
     */
    private static void Solve(int m = 3, int n = 6, int minimize_val = 0)
    {
        Solver solver = new Solver("Nontransitive_dice");

        Console.WriteLine("Number of dice: {0}", m);
        Console.WriteLine("Number of sides: {0}", n);
        Console.WriteLine("minimize_val: {0}\n", minimize_val);

        //
        // Decision variables
        //

        // The dice
        IntVar[,] dice = solver.MakeIntVarMatrix(m, n, 1, n * 2, "dice");
        IntVar[] dice_flat = dice.Flatten();

        // For comparison (probability)
        IntVar[,] comp = solver.MakeIntVarMatrix(m, 2, 0, n * n, "dice");
        IntVar[] comp_flat = comp.Flatten();

        // For branching
        IntVar[] all = dice_flat.Concat(comp_flat).ToArray();

        // The following variables are for summaries or objectives
        IntVar[] gap = solver.MakeIntVarArray(m, 0, n * n, "gap");
        IntVar gap_sum = gap.Sum().Var();

        IntVar max_val = dice_flat.Max().Var();
        IntVar max_win = comp_flat.Max().Var();

        // number of occurrences of each value of the dice
        IntVar[] counts = solver.MakeIntVarArray(n * 2 + 1, 0, n * m, "counts");

        //
        // Constraints
        //

        // Number of occurrences for each number
        solver.Add(dice_flat.Distribute(counts));

        // Order of the number of each die, lowest first
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n - 1; j++)
            {
                solver.Add(dice[i, j] <= dice[i, j + 1]);
            }
        }

        // Nontransitivity
        for (int i = 0; i < m; i++)
        {
            solver.Add(comp[i, 0] > comp[i, 1]);
        }

        // Probability gap
        for (int i = 0; i < m; i++)
        {
            solver.Add(gap[i] == comp[i, 0] - comp[i, 1]);
            solver.Add(gap[i] > 0);
        }

        // And now we roll...
        // comp[] is the number of wins for [A vs B, B vs A]
        for (int d = 0; d < m; d++)
        {
            IntVar sum1 = (from r1 in Enumerable.Range(0, n) from r2 in Enumerable.Range(0, n)
                               select(dice[d % m, r1] > dice[(d + 1) % m, r2]))
                              .ToArray()
                              .Sum()
                              .Var();

            solver.Add(comp[d % m, 0] == sum1);

            IntVar sum2 = (from r1 in Enumerable.Range(0, n) from r2 in Enumerable.Range(0, n)
                               select(dice[(d + 1) % m, r1] > dice[d % m, r2]))
                              .ToArray()
                              .Sum()
                              .Var();

            solver.Add(comp[d % m, 1] == sum2);
        }

        //
        // Search
        //
        DecisionBuilder db = solver.MakePhase(all, Solver.INT_VAR_DEFAULT, Solver.ASSIGN_MIN_VALUE);

        if (minimize_val > 0)
        {
            Console.WriteLine("Minimizing max_val");

            OptimizeVar obj = max_val.Minimize(1);

            // Other experiments:
            // OptimizeVar obj = max_win.Maximize(1);
            // OptimizeVar obj = gap_sum.Maximize(1);

            solver.NewSearch(db, obj);
        }
        else
        {
            solver.NewSearch(db);
        }

        while (solver.NextSolution())
        {
            Console.WriteLine("gap_sum: {0}", gap_sum.Value());
            Console.WriteLine("gap: {0}",
                              (from i in Enumerable.Range(0, m) select gap[i].Value().ToString()).ToArray());
            Console.WriteLine("max_val: {0}", max_val.Value());
            Console.WriteLine("max_win: {0}", max_win.Value());
            Console.WriteLine("dice:");
            for (int i = 0; i < m; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    Console.Write(dice[i, j].Value() + " ");
                }
                Console.WriteLine();
            }
            Console.WriteLine("comp:");
            for (int i = 0; i < m; i++)
            {
                for (int j = 0; j < 2; j++)
                {
                    Console.Write(comp[i, j].Value() + " ");
                }
                Console.WriteLine();
            }
            Console.WriteLine("counts:");
            for (int i = 1; i < n * 2 + 1; i++)
            {
                int c = (int)counts[i].Value();
                if (c > 0)
                {
                    Console.Write("{0}({1}) ", i, c);
                }
            }
            Console.WriteLine("\n");
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        int m = 3;            // number of dice
        int n = 6;            // number of sides of each die
        int minimize_val = 0; // minimizing max_max (0: no, 1: yes)

        if (args.Length > 0)
        {
            m = Convert.ToInt32(args[0]);
        }

        if (args.Length > 1)
        {
            n = Convert.ToInt32(args[1]);
        }

        if (args.Length > 2)
        {
            minimize_val = Convert.ToInt32(args[2]);
        }

        Solve(m, n, minimize_val);
    }
}

File: examples/contrib/nontransitive_dice.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Nontransitive dice in Google CP Solver.

 From
 http://en.wikipedia.org/wiki/Nontransitive_dice
 '''
 A set of nontransitive dice is a set of dice for which the relation
 'is more likely to roll a higher number' is not transitive. See also
 intransitivity.

 This situation is similar to that in the game Rock, Paper, Scissors,
 in which each element has an advantage over one choice and a
 disadvantage to the other.
 '''

 I start with the 3 dice version
 '''
    * die A has sides {2,2,4,4,9,9},
    * die B has sides {1,1,6,6,8,8}, and
    * die C has sides {3,3,5,5,7,7}.
 '''

 3 dice:
 Maximum winning: 27
 comp: [19, 27, 19]
 dice:
 [[0, 0, 3, 6, 6, 6],
 [2, 5, 5, 5, 5, 5],
 [1, 1, 4, 4, 4, 7]]
 max_win: 27

 Number of solutions:  1
 Nodes: 1649873  Time: 25.94
 getFailures: 1649853
 getBacktracks: 1649873
 getPropags: 98105090

Max winnings where they are the same: 21
  comp: [21, 21, 21]
  dice:
  [[0, 0, 3, 3, 3, 6],
  [2, 2, 2, 2, 2, 5],
  [1, 1, 1, 4, 4, 4]]
  max_win: 21

  Compare with these models:
  * MiniZinc: http://hakank.org/minizinc/nontransitive_dice.mzn
  * Comet: http://hakank.org/comet/nontransitive_dice.co


 This model was created by Hakan Kjellerstrand (hakank@gmail.com)
 Also see my other Google CP Solver models:
 http://www.hakank.org/google_or_tools/
"""

import sys

from ortools.constraint_solver.python import constraint_solver as cp


def main(m=3, n=6, minimize_val=0):

    # Create the solver.
    solver = cp.Solver("Nontransitive dice")

    #
    # data
    #
    print("number of dice:", m)
    print("number of sides:", n)

    #
    # declare variables
    #

    dice = {}
    for i in range(m):
        for j in range(n):
            dice[(i, j)] = solver.new_int_var(1, n * 2, "dice(%i,%i)" % (i, j))
    dice_flat = [dice[(i, j)] for i in range(m) for j in range(n)]

    comp = {}
    for i in range(m):
        for j in range(2):
            comp[(i, j)] = solver.new_int_var(0, n * n, "comp(%i,%i)" % (i, j))
    comp_flat = [comp[(i, j)] for i in range(m) for j in range(2)]

    # The following variables are for summaries or objectives
    gap = [solver.new_int_var(0, n * n, "gap(%i)" % i) for i in range(m)]
    gap_sum = solver.new_int_var(0, m * n * n, "gap_sum")

    max_val = solver.new_int_var(0, n * 2, "max_val")
    max_win = solver.new_int_var(0, n * n, "max_win")

    # number of occurrences of each value of the dice
    counts = [solver.new_int_var(0, n * m, "counts(%i)" % i) for i in range(n * 2 + 1)]

    #
    # constraints
    #

    # number of occurrences for each number
    solver.add_distribute(dice_flat, list(range(n * 2 + 1)), counts)

    solver.add(max_win == solver.max(comp_flat))
    solver.add(max_val == solver.max(dice_flat))

    # order of the number of each die, lowest first
    [
        solver.add(dice[(i, j)] <= dice[(i, j + 1)])
        for i in range(m)
        for j in range(n - 1)
    ]

    # nontransitivity
    [comp[i, 0] > comp[i, 1] for i in range(m)],

    # probability gap
    [solver.add(gap[i] == comp[i, 0] - comp[i, 1]) for i in range(m)]
    [solver.add(gap[i] > 0) for i in range(m)]
    solver.add(gap_sum == solver.sum(gap))

    # and now we roll...
    #  Number of wins for [A vs B, B vs A]
    for d in range(m):
        b1 = [
            solver.add_is_greater_var(dice[d % m, r1], dice[(d + 1) % m, r2])
            for r1 in range(n)
            for r2 in range(n)
        ]
        solver.add(comp[d % m, 0] == solver.sum(b1))

        b2 = [
            solver.add_is_greater_var(dice[(d + 1) % m, r1], dice[d % m, r2])
            for r1 in range(n)
            for r2 in range(n)
        ]
        solver.add(comp[d % m, 1] == solver.sum(b2))

    # objective
    if minimize_val != 0:
        print("Minimizing max_val")
        objective = solver.minimize(max_val, 1)
        # other experiments
        # objective = solver.maximize(max_win, 1)
        # objective = solver.maximize(gap_sum, 1)

    #
    # solution and search
    #
    db = solver.phase(
        dice_flat + comp_flat,
        cp.IntVarStrategy.INT_VAR_DEFAULT,
        cp.IntValueStrategy.ASSIGN_MIN_VALUE,
    )

    if minimize_val:
        solver.new_search(db, [objective])
    else:
        solver.new_search(db)

    num_solutions = 0
    while solver.next_solution():
        print("gap_sum:", gap_sum.value())
        print("gap:", [gap[i].value() for i in range(m)])
        print("max_val:", max_val.value())
        print("max_win:", max_win.value())
        print("dice:")
        for i in range(m):
            for j in range(n):
                print(dice[(i, j)].value(), end=" ")
            print()
        print("comp:")
        for i in range(m):
            for j in range(2):
                print(comp[(i, j)].value(), end=" ")
            print()
        print("counts:", [counts[i].value() for i in range(n * 2 + 1)])
        print()

        num_solutions += 1

    solver.end_search()

    print()
    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


m = 3  # number of dice
n = 6  # number of sides of each die
minimize_val = 0  # Minimizing max value (0: no, 1: yes)
if __name__ == "__main__":
    if len(sys.argv) > 1:
        m = int(sys.argv[1])
    if len(sys.argv) > 2:
        n = int(sys.argv[2])
    if len(sys.argv) > 3:
        minimize_val = int(sys.argv[3])

    main(m, n, minimize_val)
